geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroSchemas.scala

```diff
@@ -0,0 +1,253 @@
+
+    private[this] def computeType(s:Schema, nillable:Boolean, multi: Boolean):Schema = {
+      val maybeMulti = if(multi) Schema.createArray(s) else s
+      if(nillable) Schema.createUnion(List(Schema.create(Schema.Type.NULL), maybeMulti)) else maybeMulti
+    }
+
+    private[this] def computeFields[R](parent: ComplexType): Iterable[Field] = {
+      allProps(parent).map(_.asInstanceOf[AttributeDescriptor]).flatMap { ad=>
+        ad.getType match {
+          case ct: ComplexType =>
+            Some(computeField(parent, ad.getName, computeType(this(ct), ad)))
+          case _ if ad.isList =>
+            for(clazz<-ad.getListType(); codec<-registry.find(clazz)) yield
+              computeField(parent, ad.getName, computeType(Schema.createArray(codec.schema), ad))
+          case _ if ad.isMap =>
+            for((keyClass,valClass)<-ad.getMapTypes();
+                keyCodec<-registry.find(keyClass);
+                valCodec<-registry.find(valClass)) yield {
+              val keyField = new Field("k", keyCodec.schema, null, null)
+              val valField = new Field("v", valCodec.schema, null, null)
+              val schema = Schema.createRecord(List(keyField,valField))
+              computeField(parent, ad.getName, computeType(Schema.createArray(schema), ad))
+            }
+          case _ =>
+            val st = nearestSimpleType(ad.getType)
+            for(codec<-registry.find(st.getBinding)) yield computeField(parent, ad.getName, computeType(codec.schema, ad))
+        }
+      }
+    }
+  }
+
+  def subclassByRestriction(child: PropertyType) =
+    isDescendedFrom(child, GMLSchema.FEATUREPROPERTYTYPE_TYPE)
+
+  def allProps(ct:ComplexType):Iterable[PropertyDescriptor] = ct.getSuper match {
+    case st:ComplexType if subclassByRestriction(st) => allProps(st).drop(ct.getDescriptors.size())++ct.getDescriptors
+    case st:ComplexType => allProps(st)++ct.getDescriptors
+    case _ => ct.getDescriptors
+  }
+
+  def isDescendedFrom(child: PropertyType, ancestor: PropertyType): Boolean =
+    Iterator.iterate(child)(_.getSuper).takeWhile(_ != null).exists(_.eq(ancestor))
+
+  @scala.annotation.tailrec
+  def nearestSimpleType(at: PropertyType): PropertyType = at match {
+    case ct:ComplexType => nearestSimpleType(ct.getSuper)
+    case st => st
+  }
+
+  sealed trait SchemaType {
+    def skip(d:Decoder)
+  }
+
+  final case class Nullable(schema: SchemaType) extends SchemaType {
+    def this(schema: SchemaType, nullidx:Int) = { this(schema); this.nullidx = nullidx; }
+    private[this] var nullidx: Int = 0
+    def writeNull(e:Encoder) = { e.writeIndex(nullidx); e.writeNull(); }
+    def writeNonNull(e:Encoder) = e.writeIndex(1 - nullidx)
+    def readNull(d:Decoder) = if(d.readIndex() == nullidx) { d.readNull(); true; } else false
+    def skip(d:Decoder) = if(!readNull(d)) schema.skip(d)
+  }
+
+  case class Array(schema: SchemaType) extends SchemaType {

```

Comment:
For the implementations of SchemaType, I'd ask for longer class names.

In particular, 'Array' is a no-go as a class name;).